keps/sig-auth/20190607-certificates-api.md

```diff
@@ -132,6 +134,85 @@ A typical successful issuance proceeds as follows.
+    May be auto-approved by kube-controller-manager.
+    1. Trust distribution: signed certificates must be honored as client-certificates by the kube-apiserver.  The CA bundle
+       is not distributed by any other means.
+    2. Permitted subjects - organizations are exactly `[]string{"system:nodes"}`, common name starts with `"system:node:"`
+    3. Permitted x509 extensions - none
+    4. Permitted key usages - exactly `[]string{"key encipherment", "digital signature", "client auth"}`
+    5. Expiration/cert lifetime - minimum of CSR signer or request.  Sanity of the time is the concern of the signer.
+    6. CA bit allowed/disallowed - not allowed.
+ 3. kubernetes.io/kubelet-serving - signs serving certificates that are honored as a valid kubelet serving certificate 
+    by the kube-apiserver, but has no other guarantees.  Never auto-approved by kube-controller-manager.
+    1. Trust distribution: signed certificates must be honored by the kube-apiserver as valid to terminate connections to a kubelet.
+       The CA bundle is not distributed by any other means.
+    2. Permitted subjects - organizations are exactly `[]string{"system:nodes"}`, common name starts with `"system:node:"`
+    3. Permitted x509 extensions - SAN DNS has value
+    4. Permitted key usages - exactly `[]string{"key encipherment", "digital signature", "server auth"}`
+    5. Expiration/cert lifetime - minimum of CSR signer or request.
+    6. CA bit allowed/disallowed - not allowed.
+ 4. kubernetes.io/legacy-unknown - has no guarantees for trust at all.  Some distributions may honor these as client
+    certs, but that behavior is not standard kubernetes behavior.  Never auto-approved by kube-controller-manager.
+    1. Trust distribution: None.  There is no standard trust or distribution for this signer in a kubernetes cluster.
+    2. Permitted subjects - any
+    3. Permitted x509 extensions - honors SAN extensions and discards other extensions.
+    4. Permitted key usages - may include `[]string{"key encipherment", "digital signature", "client auth", "server auth"}`
+    5. Expiration/cert lifetime - minimum of CSR signer or request.  Sanity of the time is the concern of the signer.
+    6. CA bit allowed/disallowed - not allowed.
+
+Distribution of trust happens out of band for these signers.  Any trust outside of those described above are strictly
+coincidental.  For instance, some distributions may honor kubernetes.io/legacy-unknown as client-certificates for the
+kube-apiserver, but this is not a standard.
+None of these usages are related to ServiceAccount token secrets `.data[ca.crt]` in any way.  That ca-bundle is only
+guaranteed to verify a connection the kube-apiserver using the default service.
+
+To support HA upgrades, the kube-controller-manager will duplicate defaulting code for an empty `signerName` for one
+release.
+
+#### Limiting approval powers for certain signers.
```

Comment:
What's the compatibility story of v1beta1? Do we skip this check for the legacy signer?

Reply:
> What's the compatibility story of v1beta1? Do we skip this check for the legacy signer?

Similar to other cases where we have chosen to tighten security, we can create a role that cluster-admins can bind as they see fit.

Reply:
Policy additions outside of bootstrappolicy are annoying for HA clusters but not terribly so. The harder part will be to make sure people know that they need migrate. I'll think through the pros and cons of skipping the check for the legacy signer. Off the cuff though:

- \+ we don't break compatibility of v1beta1
- \+ we already have a date(-ish) for removal of the legacy signer
- \+ we bundle all the migration work together into the removal of the v1beta1 API
- \- the date for removal of legacy signer still feels very away
- \- during that time, the legacy signer will be very powerful in many clusters, often more powerful than other other signers. these clusters won't benefit from the improved security of signer isolation.

Reply:
Discussed in sig-auth. Decisions:
* it is better to be consistent with behavior among signer names, to contain behavioral changes to the beta phase, and to not exempt the legacy signer (especially since it is likely a very powerful signer)
* we will consistently run signer-specific authorization checks on approval in an admission plugin
* we will provide 3 bootstrap cluster-roles defining authorization rules needed to approve CSRs for the kubernetes.io signerNames
* cluster admins can either:
  * grant signer-specific approval permissions using roles they define
  * grant signer-specific approval permissions using the bootstrap roles starting in 1.18
  * disable the approval-authorizing admission plugin in 1.18 (if they don't care about partitioning approver rights)